packages/preview/src/browser/preview-widget-manager.ts

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright (C) 2018 TypeFox and others.
+ *
+ * Licensed under the Apache License, Version 2.0 (the 'License'); you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+import { interfaces, injectable } from 'inversify';
+import { DisposableCollection } from '@theia/core';
+import URI from "@theia/core/lib/common/uri";
+import { WidgetFactory } from '@theia/core/lib/browser';
+import { Emitter, Event, } from '@theia/core/lib/common';
+import { PreviewWidget, PREVIEW_WIDGET_FACTORY_ID } from './preview-widget';
+
+@injectable()
+export class PreviewWidgetManager implements WidgetFactory {
+
+    readonly id: string = PREVIEW_WIDGET_FACTORY_ID;
+
+    protected readonly onWidgetCreatedEmitter = new Emitter<string>();
+
+    protected readonly disposables = new DisposableCollection();
+    private widgets = new Map<string, PreviewWidget>();
```

Comment:
you don't need to maintain your own cache. The `WidgetManager`does that already.

Reply:
the main purpose of `PreviewWidgetManager` is to provide `onWidgetCreated` event, which is used to update synchronization with a maybe existing editor widget. 

unfortunately the `WidgetManager` will add that widget a step after calling `createWidget` on the factory. but it's possible to refactor this, and to notify with the created widget instead of uri.